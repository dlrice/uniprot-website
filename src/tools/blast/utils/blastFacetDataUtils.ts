import {
  getSmallerMultiple,
  getLargerMultiple,
} from '../../../shared/utils/utils';

import { BlastResults, BlastFacets } from '../types/blastResults';

export const isBlastValueWithinRange = (
  blastResultDatum,
  rangeFilters,
  facet
) => {
  try {
    const value = blastResultDatum[facet];
    const [min, max] = rangeFilters[facet];
    return min <= value && value <= max;
  } catch (e) {
    console.log('error:', e);
  }
};

export const filterBlastDatum = (
  blastResultDatum,
  rangeFilters,
  activeFacet
) => {
  // All inactiveFacets (including user selected and not user selected) will need to have the intersection of all of the ranges applied (including the active).
  // The activeFacet has only has the inactiveFacets intersection applied.

  // TODO check if the ranges aren't empty arrays

  if (!rangeFilters || !Object.keys(rangeFilters).length) {
    return blastResultDatum;
  }

  const inactiveRangedFacets = Object.keys(rangeFilters).filter(
    (facet) => facet !== activeFacet
  );

  const includeActive = inactiveRangedFacets.every((facet) =>
    isBlastValueWithinRange(blastResultDatum, rangeFilters, facet)
  );
  const includeInactive =
    includeActive &&
    isBlastValueWithinRange(blastResultDatum, rangeFilters, activeFacet);

  const result = {};

  const allInactiveFacets = Object.values(BlastFacets).filter(
    (facet) => facet !== activeFacet
  );

  allInactiveFacets.forEach((facet) => {
    if (includeInactive) {
      result[facet] = blastResultDatum[facet];
    }
  });

  if (includeActive) {
    result[activeFacet] = blastResultDatum[activeFacet];
  }

  return result;
};

const setMinMaxValues = (results, facet, value) => {
  if (typeof results[facet].min === 'undefined' || results[facet].min > value) {
    results[facet].min = value;
  }

  if (typeof results[facet].max === 'undefined' || results[facet].max < value) {
    results[facet].max = value;
  }
};

export const getBlastParametersFacetsFromData = (
  facets: SelectedFacet[],
  activeFacet: string,
  data?: BlastResults | null,
  histogramBinSize: number
) => {
  const results = {
    score: {
      values: [],
      min: undefined,
      max: undefined,
    },
    identity: {
      values: [],
      min: undefined,
      max: undefined,
    },
    evalue: {
      values: [],
      min: undefined,
      max: undefined,
    },
  };

  if (!data) {
    return results;
  }

  const parsedFacets = Object.fromEntries(
    facets.map(({ name, value }) => {
      const [min, max] = value.split('-').map((x) => Number(x));
      return [name, [min, max]];
    })
  );

  data.hits.forEach(({ hit_hsps }) => {
    hit_hsps.forEach(({ hsp_score, hsp_identity, hsp_expect }) => {
      const datum = {
        score: hsp_score,
        identity: hsp_identity,
        evalue: hsp_expect,
      };

      const { score, identity, evalue } = filterBlastDatum(
        datum,
        parsedFacets,
        activeFacet
      );

      setMinMaxValues(results, 'score', hsp_score);
      setMinMaxValues(results, 'identity', hsp_identity);
      setMinMaxValues(results, 'evalue', hsp_expect);

      // We would like to include 0 values, hence, check for 'undefined' explicitly
      if (score !== undefined) {
        results.score.values.push(score);
      }
      if (identity !== undefined) {
        results.identity.values.push(identity);
      }
      if (evalue !== undefined) {
        results.evalue.values.push(evalue);
      }
    });
  });
  console.log('results:', results);
  return results;
};
