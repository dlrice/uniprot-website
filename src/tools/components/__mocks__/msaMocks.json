{
  "BLAST": [
    {
      "name": "Query",
      "sequence": "MKRTVISSSNAYASKRSRLDIEHDFEQYHSLNKKYYPRPITRTGANQFNNKSRAKPMEIVEKLQKKQKTSFENVSTVMHWFRNDLRLYDNVGLYKSVALFQQLRQKNAKAKLYAVYVINEDDWRAHMDSGWKLMFIMGALKNLQQSLAELHIPLLLWEFHTPKSTLSNSKEFVEFFKEKCMNVSSGTGTIITANIEYQTDELYRDIRLLENEDHRLQLKYYHDSCIVAPGLITTDRGTNYSVFTPWYKKWVLYVNNYKKSTSEICHLHIIEPLKYNETFELKPFQYSLPDEFLQYIPKSKWCLPDVSEEAALSRLKDFLGTKSSKYNNEKDMLYLGGTSGLSVYITTGRISTRLIVNQAFQSCNGQIMSKALKDNSSTQNFIKEVAWRDFYRHCMCNWPYTSMGMPYRLDTLDIKWENNPVAFEKWCTGNTGIPIVDAIMRKLLYTGYINNRSRMITASFLSKNLLIDWRWGERWFMKHLIDGDSSSNVGGWGFCSSTGIDAQPYFRVFNMDIQAKKYDPQMIFVKQWVPELISSENKRPENYPKPLVDLKHSRERALKVYKDAM",
      "from": 1,
      "to": 565,
      "length": 565
    },
    {
      "name": "Match:P05066",
      "sequence": "MKRTVISSSNAYASKRSRLDIEHDFEQYHSLNKKYYPRPITRTGANQFNNKSRAKPMEIVEKLQKKQKTSFENVSTVMHWFRNDLRLYDNVGLYKSVALFQQLRQKNAKAKLYAVYVINEDDWRAHMDSGWKLMFIMGALKNLQQSLAELHIPLLLWEFHTPKSTLSNSKEFVEFFKEKCMNVSSGTGTIITANIEYQTDELYRDIRLLENEDHRLQLKYYHDSCIVAPGLITTDRGTNYSVFTPWYKKWVLYVNNYKKSTSEICHLHIIEPLKYNETFELKPFQYSLPDEFLQYIPKSKWCLPDVSEEAALSRLKDFLGTKSSKYNNEKDMLYLGGTSGLSVYITTGRISTRLIVNQAFQSCNGQIMSKALKDNSSTQNFIKEVAWRDFYRHCMCNWPYTSMGMPYRLDTLDIKWENNPVAFEKWCTGNTGIPIVDAIMRKLLYTGYINNRSRMITASFLSKNLLIDWRWGERWFMKHLIDGDSSSNVGGWGFCSSTGIDAQPYFRVFNMDIQAKKYDPQMIFVKQWVPELISSENKRPENYPKPLVDLKHSRERALKVYKDAM",
      "from": 1,
      "to": 565,
      "length": 565,
      "accession": "P05066",
      "features": [
        {
          "protvistaFeatureId": "cb1e3e80-227f-11eb-8c7f-bd249b5a2490",
          "start": 75,
          "end": 226,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "Other domain of interest",
          "description": "Photolyase/cryptochrome alpha/beta"
        },
        {
          "protvistaFeatureId": "cb1e3e81-227f-11eb-8c7f-bd249b5a2490",
          "start": 338,
          "end": 342,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "Nucleotide-binding region",
          "description": "FAD",
          "evidences": [
            {
              "evidenceCode": "ECO:0000250"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e82-227f-11eb-8c7f-bd249b5a2490",
          "start": 482,
          "end": 484,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "Nucleotide-binding region",
          "description": "FAD",
          "evidences": [
            {
              "evidenceCode": "ECO:0000250"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e83-227f-11eb-8c7f-bd249b5a2490",
          "start": 384,
          "end": 391,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "region of interest",
          "description": "Interaction with DNA",
          "evidences": [
            {
              "evidenceCode": "ECO:0000250"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e84-227f-11eb-8c7f-bd249b5a2490",
          "start": 451,
          "end": 452,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "region of interest",
          "description": "Interaction with DNA",
          "evidences": [
            {
              "evidenceCode": "ECO:0000250"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e85-227f-11eb-8c7f-bd249b5a2490",
          "start": 326,
          "end": 326,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "Other binding site",
          "description": "FAD",
          "evidences": [
            {
              "evidenceCode": "ECO:0000250"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e86-227f-11eb-8c7f-bd249b5a2490",
          "start": 514,
          "end": 514,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "Other binding site",
          "description": "DNA",
          "evidences": [
            {
              "evidenceCode": "ECO:0000250"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e87-227f-11eb-8c7f-bd249b5a2490",
          "start": 416,
          "end": 416,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "Other site of interest",
          "description": "Electron transfer via tryptophanyl radical",
          "evidences": [
            {
              "evidenceCode": "ECO:0000250"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e88-227f-11eb-8c7f-bd249b5a2490",
          "start": 469,
          "end": 469,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "Other site of interest",
          "description": "Electron transfer via tryptophanyl radical",
          "evidences": [
            {
              "evidenceCode": "ECO:0000250"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e89-227f-11eb-8c7f-bd249b5a2490",
          "start": 492,
          "end": 492,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "Other site of interest",
          "description": "Electron transfer via tryptophanyl radical",
          "evidences": [
            {
              "evidenceCode": "ECO:0000250"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e8a-227f-11eb-8c7f-bd249b5a2490",
          "start": 387,
          "end": 387,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "mutagenesis site",
          "description": "Reduces substrate binding 100-fold. Reduces quantum yield for dimer photolysis 3-fold.",
          "evidences": [
            {
              "evidenceCode": "ECO:0000269",
              "source": "PubMed",
              "id": "8344951"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e8b-227f-11eb-8c7f-bd249b5a2490",
          "start": 463,
          "end": 463,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "mutagenesis site",
          "description": "Reduces substrate binding 100-fold.",
          "evidences": [
            {
              "evidenceCode": "ECO:0000269",
              "source": "PubMed",
              "id": "8344951"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e8c-227f-11eb-8c7f-bd249b5a2490",
          "start": 507,
          "end": 507,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "mutagenesis site",
          "description": "Reduces substrate binding 100-fold.",
          "evidences": [
            {
              "evidenceCode": "ECO:0000269",
              "source": "PubMed",
              "id": "8344951"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e8d-227f-11eb-8c7f-bd249b5a2490",
          "start": 517,
          "end": 517,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "mutagenesis site",
          "description": "Reduces substrate binding 10-fold."
        },
        {
          "protvistaFeatureId": "cb1e3e8e-227f-11eb-8c7f-bd249b5a2490",
          "start": 77,
          "end": 77,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "sequence conflict",
          "description": "in Ref. 2; AAA34875",
          "evidences": [
            {
              "evidenceCode": "ECO:0000305"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e8f-227f-11eb-8c7f-bd249b5a2490",
          "start": 165,
          "end": 165,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "sequence conflict",
          "description": "in Ref. 2; AAA34875",
          "evidences": [
            {
              "evidenceCode": "ECO:0000305"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e90-227f-11eb-8c7f-bd249b5a2490",
          "start": 169,
          "end": 169,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "sequence conflict",
          "description": "in Ref. 2; AAA34875",
          "evidences": [
            {
              "evidenceCode": "ECO:0000305"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e91-227f-11eb-8c7f-bd249b5a2490",
          "start": 200,
          "end": 200,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "sequence conflict",
          "description": "in Ref. 2; AAA34875",
          "evidences": [
            {
              "evidenceCode": "ECO:0000305"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e92-227f-11eb-8c7f-bd249b5a2490",
          "start": 351,
          "end": 351,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "sequence conflict",
          "description": "in Ref. 2; AAA34875",
          "evidences": [
            {
              "evidenceCode": "ECO:0000305"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e93-227f-11eb-8c7f-bd249b5a2490",
          "start": 365,
          "end": 365,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "sequence conflict",
          "description": "in Ref. 2; AAA34875",
          "evidences": [
            {
              "evidenceCode": "ECO:0000305"
            }
          ]
        },
        {
          "protvistaFeatureId": "cb1e3e94-227f-11eb-8c7f-bd249b5a2490",
          "start": 473,
          "end": 473,
          "startModifier": "EXACT",
          "endModifier": "EXACT",
          "type": "sequence conflict",
          "description": "in Ref. 2; AAA34875",
          "evidences": [
            {
              "evidenceCode": "ECO:0000305"
            }
          ]
        }
      ]
    }
  ]
}
