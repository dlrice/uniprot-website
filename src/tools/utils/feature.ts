import { cloneDeep } from 'lodash-es';
import { processUrlTemplate } from '../../uniprotkb/components/protein-data-views/XRefView';
import evidenceUrls from '../../uniprotkb/config/evidenceUrls';

export const prepareFeatureForTooltip = (feature) => {
  const cloned = cloneDeep(feature);

  if (feature.featureId) {
    cloned.ftId = feature.featureId;
  }

  if (!feature.evidences) {
    return cloned;
  }

  cloned.evidences = cloned.evidences.map((e) => {
    const prepared = { ...e, code: e.evidenceCode };
    if (e.id && e.source) {
      const source = {
        id: e.id,
        name: e.source,
      };
      const url = evidenceUrls[e.source];
      if (url) {
        source.url = processUrlTemplate(url, {
          value: e.id,
        });
      }
      if (e.source === 'PubMed') {
        source.alternativeUrl = processUrlTemplate(evidenceUrls.EuropePMC, {
          value: e.id,
        });
      }
    }
    return prepared;
  });
  return cloned;
};
