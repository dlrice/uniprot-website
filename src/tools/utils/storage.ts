import { Store, set, get, del, clear, keys } from 'idb-keyval';

import { Stores } from './stores';

export const DB_NAME = 'UniProt-Jobs';

export default class JobStore {
  #store: Store;

  constructor(storeName: Stores) {
    this.#store = new Store(DB_NAME, storeName);
  }

  async get(key: IDBValidKey) {
    return get(key, this.#store);
  }

  async *getAll() {
    const allKeys = await this.keys();
    for (const key of allKeys) {
      yield this.get(key);
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  async set(key: IDBValidKey, value: any) {
    // TODO: uncomment when we have finished coding
    // return set(key, value, this.#store);
  }

  async del(key: IDBValidKey) {
    return del(key, this.#store);
  }

  async clear() {
    return clear(this.#store);
  }

  async keys() {
    return keys(this.#store);
  }
}
