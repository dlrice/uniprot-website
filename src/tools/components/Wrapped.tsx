/* eslint-disable camelcase */
/* eslint-disable no-param-reassign */
import React, {
  FC,
  useCallback,
  useMemo,
  useRef,
  SetStateAction,
  Dispatch,
} from 'react';
import { debounce } from 'lodash-es';
import ProtvistaManager from 'protvista-manager';
import ProtvistaTrack from 'protvista-track';
import ProtvistaMSA from 'protvista-msa';

import { loadWebComponent } from '../../shared/utils/utils';

import useSize from '../../shared/hooks/useSize';
import useSafeState from '../../shared/hooks/useSafeState';
import useStaggeredRenderingHelper from '../../shared/hooks/useStaggeredRenderingHelper';

import { MsaColorScheme } from '../config/msaColorSchemes';

import FeatureType from '../../uniprotkb/types/featureType';
import { ConservationOptions, MSAInput } from './AlignmentView';
import {
  FeatureData,
  processFeaturesData,
} from '../../uniprotkb/components/protein-data-views/FeaturesView';
import {
  transformFeaturesPositions,
  getEndCoordinate,
} from '../utils/sequences';
import AlignLabel from '../align/components/results/AlignLabel';

loadWebComponent('protvista-track', ProtvistaTrack);
loadWebComponent('protvista-msa', ProtvistaMSA);
loadWebComponent('protvista-manager', ProtvistaManager);

const widthOfAA = 20;

export type Sequence = {
  name: string;
  sequence: string;
  start: number;
  end: number;
  features?: FeatureData;
  accession?: string;
};

type Chunk = {
  id: string;
  sequences: Sequence[];
};

export type MSAWrappedRowProps = {
  rowLength: number;
  highlightProperty: MsaColorScheme | undefined;
  conservationOptions: ConservationOptions;
  annotation: FeatureType | undefined;
  sequences: Sequence[];
  activeId?: string;
  setActiveId?: Dispatch<SetStateAction<string | undefined>>;
  omitInsertionsInCoords?: boolean;
  selectedEntries?: string[];
  handleSelectedEntries?: (rowId: string) => void;
};

// NOTE: hardcoded for now, might need to change that in the future if need be
const heightStyle = { height: '20px' };

const MSAWrappedRow: FC<MSAWrappedRowProps> = ({
  rowLength,
  highlightProperty,
  conservationOptions,
  annotation,
  sequences,
  activeId,
  setActiveId,
  omitInsertionsInCoords = true,
  selectedEntries,
  handleSelectedEntries,
}) => {
  const setMSAAttributes = useCallback(
    (node): void => {
      if (!node) {
        return;
      }
      node.data = sequences;
    },
    [sequences]
  );

  const setFeatureTrackData = useCallback(
    (node): void => {
      if (node && annotation) {
        const featuresSeq = sequences.find(
          ({ accession }) => accession && accession === activeId
        );
        const features = featuresSeq?.features?.filter(
          ({ type }) => type === annotation
        );
        if (featuresSeq && features) {
          let processedFeatures = processFeaturesData(features);
          processedFeatures = transformFeaturesPositions(processedFeatures);
          node.data = processedFeatures;
          node.setAttribute('length', featuresSeq.end - featuresSeq.start + 1);
          node.setAttribute('displaystart', featuresSeq.start);
          node.setAttribute('displayend', featuresSeq.end);
        }
      }
    },
    [annotation, activeId, sequences]
  );

  // TODO: replace this with fragments to have one big grid
  // -> to keep the right column of the right size to fit all possible values
  return (
    <>
      <div className="track-label track-label--align-labels">
        {sequences.map((s) => (
          <AlignLabel
            accession={s.accession}
            info={s}
            loading={false}
            key={s.name}
            style={heightStyle}
            checked={Boolean(
              s.accession && selectedEntries?.includes(s.accession)
            )}
            onSequenceChecked={handleSelectedEntries}
            onIdClick={() => setActiveId?.(s.accession)}
            active={!!activeId && setActiveId && activeId === s.accession}
          >
            {s.name || ''}
          </AlignLabel>
        ))}
      </div>
      <div className="track">
        <protvista-msa
          ref={setMSAAttributes}
          length={rowLength}
          height={sequences.length * 20}
          colorscheme={highlightProperty}
          {...conservationOptions}
        />
      </div>
      <span className="right-coord">
        {sequences.map((s) => (
          <div style={heightStyle} key={s.name}>
            {omitInsertionsInCoords
              ? getEndCoordinate(s.sequence, s.end)
              : s.end}
          </div>
        ))}
      </span>
      <span className="track-label annotation-label">{annotation}</span>
      <div className="track annotation-track">
        <protvista-track ref={setFeatureTrackData} />
      </div>
    </>
  );
};

export type MSAViewProps = {
  alignment: MSAInput[];
  alignmentLength: number;
  highlightProperty: MsaColorScheme | undefined;
  conservationOptions: ConservationOptions;
  totalLength: number;
  annotation: FeatureType | undefined;
  activeId?: string;
  setActiveId?: Dispatch<SetStateAction<string | undefined>>;
  omitInsertionsInCoords?: boolean;
  selectedEntries?: string[];
  handleSelectedEntries?: (rowId: string) => void;
};

const Wrapped: FC<MSAViewProps> = ({
  alignment,
  alignmentLength,
  highlightProperty,
  conservationOptions,
  annotation,
  activeId,
  setActiveId,
  omitInsertionsInCoords = false,
  selectedEntries,
  handleSelectedEntries,
}) => {
  const containerRef = useRef<HTMLDivElement>(null);
  const [size] = useSize(containerRef);

  const [rowLength, setRowLength] = useSafeState(0);
  const nItemsToRender = useStaggeredRenderingHelper({
    first: 4,
    increment: +Infinity,
    max: +Infinity,
    delay: 500,
  });

  const debouncedSetRowLength = useMemo(
    () =>
      debounce((width: number) => {
        // using 9 tenths of the available size as its the proportion assigned
        // to the track in the CSS
        setRowLength(Math.floor((0.9 * width) / widthOfAA));
      }, 1000),
    [setRowLength]
  );

  if (size?.width) {
    debouncedSetRowLength(size.width);
    if (!rowLength) {
      // when it passes from 0 to any value, don't debounce
      debouncedSetRowLength.flush();
    }
  }

  const sequenceChunks = useMemo<Chunk[]>(() => {
    if (!rowLength) {
      return [];
    }

    const numberRows = Math.ceil(alignmentLength / rowLength);
    const chunks = [...Array(numberRows).keys()].map((index) => {
      const start = index * rowLength;
      const end = start + rowLength;
      return {
        // NOTE: This is a bit of an ugly trick to have the whole track
        // re-render on change of size. Otherwise when the track is updated
        // instead of re-rendered, we get the track overflowing on the right.
        // Might be able to avoid that by playing with sizes in the panel grid
        // and from within the Nightingale component
        id: `row-${index}-${rowLength}`,
        sequences: alignment.map(
          ({ name, sequence, from, features, accession }) => ({
            name: name || '',
            sequence: sequence.slice(start, end),
            start: start + from,
            end: end + from - 1,
            features,
            accession,
          })
        ),
      };
    });
    return chunks;
  }, [alignment, alignmentLength, rowLength]);

  return (
    <div
      ref={containerRef}
      className="alignment-grid alignment-wrapped"
      data-testid="alignment-wrapped-view"
    >
      {sequenceChunks.map(({ sequences, id }, index) => {
        if (index < nItemsToRender) {
          return (
            <MSAWrappedRow
              key={id}
              rowLength={rowLength}
              sequences={sequences}
              annotation={annotation}
              highlightProperty={highlightProperty}
              conservationOptions={conservationOptions}
              activeId={activeId}
              setActiveId={setActiveId}
              omitInsertionsInCoords={omitInsertionsInCoords}
              selectedEntries={selectedEntries}
              handleSelectedEntries={handleSelectedEntries}
            />
          );
        }
        return <div key={id} className="alignment-grid__placeholder" />;
      })}
    </div>
  );
};

export default Wrapped;
