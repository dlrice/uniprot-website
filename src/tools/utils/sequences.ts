/* eslint-disable camelcase */
import { v1 } from 'uuid';

import { MSAInput } from '../components/AlignmentView';
import {
  ProcessedFeature,
  FeatureData,
  LocationModifier,
} from '../../uniprotkb/components/protein-data-views/FeaturesView';
import { Sequence } from '../components/Wrapped';

export const findSequenceSegments = (seq: string) => {
  const ranges: number[][] = [];
  const newRange = () => ({
    start: null,
    end: null,
  });

  type SegmentRange = {
    start: number | null;
    end: number | null;
  };

  let range: SegmentRange = newRange();

  [...seq].forEach((ch, i) => {
    if (ch !== '-') {
      if (range.start === null) {
        range.start = i + 1;
      }
    } else if (range.start !== null && range.end === null) {
      range.end = i;
      ranges.push([range.start, range.end]);
      range = newRange();
    }

    if (i === seq.length - 1 && range.start !== null && range.end === null) {
      range.end = i + 1;
      ranges.push([range.start as number, range.end as number]);
      range = newRange();
    }
  });

  return ranges;
};

export const getFullAlignmentLength = (
  alignment: MSAInput[],
  alignmentLength: number
) => {
  /**
   * prefix|align|suffix
   *     --|=====|-----
   * ------|=====|---------
   */
  const prefix = Math.max(...alignment.map(({ from }) => from));
  const suffix = Math.max(...alignment.map(({ length, to }) => length - to));
  return prefix - 1 + alignmentLength + suffix;
};

export type SegmentTrackData = {
  start: number;
  end: number;
  shape?: string;
  color: string;
};

export type FullAlignmentSegments = {
  name?: string;
  accession?: string;
  sequence: string;
  from: number;
  to: number;
  length: number;
  features?: FeatureData;
  trackData: SegmentTrackData[];
};

export const getFullAlignmentSegments = (alignment: MSAInput[]) => {
  // franklin $colour-sapphire-blue
  const colour = '#014371';

  const countGaps = (seq: string) => {
    const matches = seq.match(/-/g);
    return matches ? matches.length : 0;
  };

  const maxFrom = Math.max(...alignment.map(({ from }) => from));

  return alignment.map((al) => {
    const offset = maxFrom - al.from > 0 ? maxFrom - al.from : 0;
    return {
      ...al,
      trackData: [
        {
          start: offset,
          end: offset + al.from - 1,
          shape: 'line',
          color: colour,
        },
        ...findSequenceSegments(al.sequence).map(([start, end]) => ({
          start: offset + (al.from - 1) + start,
          end: offset + (al.from - 1) + end,
          color: colour,
        })),
        {
          start: offset + al.to + countGaps(al.sequence) + 1,
          end: offset + al.length + countGaps(al.sequence),
          shape: 'line',
          color: colour,
        },
      ],
    };
  });
};

export const transformFeaturesPositions = (features: ProcessedFeature[]) =>
  features.map((feature) => ({
    ...feature,
    start: feature.start - 1,
    end: feature.end - 1,
  }));

export const getNumberOfInsertions = (sequence: string, endPosition: number) =>
  (sequence.slice(0, endPosition).match(/-/g) || []).length;

export const getEndCoordinate = (sequence: string, endPosition: number) =>
  endPosition - getNumberOfInsertions(sequence, endPosition);

// Jie has said that if it is unknown, you can ignore value
// These erroneous features are temporary and will eventually be removed
export const removeFeaturesWithUnknownModifier = (features?: FeatureData) =>
  features?.filter(
    ({ location: { start, end } }) =>
      start.modifier !== LocationModifier.UNKNOWN &&
      end.modifier !== LocationModifier.UNKNOWN
  );

export const createGappedFeature = (feature, sequence) => {
  /*
  input: feature and sequence are both 1-based
  */
  const BLOCK = /(?<insertion>[-]+)|(?<protein>[^-]+)/g;

  let proteinIndex = 1;
  const fragments = [];

  let match;
  // eslint-disable-next-line no-cond-assign
  while ((match = BLOCK.exec(sequence)) !== null) {
    if (match?.groups?.protein) {
      if (
        proteinIndex + match[0].length >= feature.start &&
        proteinIndex <= feature.end
      ) {
        fragments.push({
          // start: match.index + 1,
          start: match.index + 1 + Math.max(0, feature.start - proteinIndex),
          end:
            match.index +
            Math.min(match[0].length, feature.end - proteinIndex + 1),
        });
      }
      proteinIndex += match[0].length;
    } else if (proteinIndex > feature.start && proteinIndex < feature.end) {
      fragments.push({
        start: match.index + 1,
        end: match.index + match[0].length,
        shape: 'line',
      });
    }
  }
  const gappedFeature = {
    ...feature,
    start: fragments[0].start,
    end: fragments[fragments.length - 1].end,
  };
  if (fragments.length > 1) {
    gappedFeature.locations = [{ fragments }];
  }
  return gappedFeature;
};

export const createGappedFeature2 = (feature, sequence /* full sequence */) => {
  const { start, end } = feature;

  // Gap positions in the full sequence
  // Format: [start, end, length of gap, start index for debugging]
  const gaps = [];
  const fragments = []; // rectangular features to represent feature blocks

  const regex = /[-]+/g;
  let match;

  // We cut off the sequence from where the feature starts before searching for
  // gaps. This is because gaps that come before the start of the feature don't
  // count, but they will mess up with our search in the sequence, if they are
  // not removed from the sequence. One should be subtracted from 'start' value
  // because string functions work on a 0-based array.
  while ((match = regex.exec(sequence.substring(start - 1))) !== null) {
    // We need to keep track of how many gaps we have encountered
    // so far in the sequence and account for those gaps in our feature
    // poistions.
    const prevoiusGaps = gaps.reduce((count, gap) => count + gap.length, 0); // number of previous gaps + length of current one

    // Recording the gap details
    gaps.push({
      // start of the gap: index - previous gaps + 1 because index starts from 0
      start: match.index - prevoiusGaps + 1,
      // end of the gap: index - previous gaps + length of current gap
      end: match.index - prevoiusGaps + match[0].length,
      // length of current gap
      length: match[0].length,
      // index is the index of the first '-' found in the sequence. this index value
      // is moved forward to the next match e.g. 2 and then 5 if sequence is '01-34-6'
      rawIndex: match.index,
    });
  }

  // Now that gaps positions and length are ready, we can move on to create the
  // fragments that represent the feature itself rather the gaps.
  //
  // Calculating start: if this the first fragment we are creating, we need to
  // take the start of the feature as the start of our fragment.
  // If this is not the first fragment that we are creating, we need to find the
  // end of previous gap -- where the previous gap ended in the sequence, and use
  // that point + 1 as our starting point.
  //
  // Example: sequence '01-34--78'. feature [0,8]. gaps[[2,2],[5,6]].
  // first fragment: start would be 0 which is the start of the feature.
  // second fragment: start would be 2, which is the end of first gap, + 1
  // thid fragment: start would be 6, whish is then eod of second gap, + 1
  //
  // Calculating end: the start of current gap -- remember we are creating the rectangular
  // feature that is behind the gap right ahead of it. Current gap -- in the loop, points to
  // the gap ahead.
  gaps.forEach((gapAhead, index) => {
    const lastGap = index === 0 ? gaps[0] : gaps[index - 1];

    // Check here if the feature hasn't already ended before overlapping with this gap
    // TODO is this right?
    // if (index !== 0 && end >= lastGap.end) {
    //   return;
    // }

    fragments.push({
      ...feature,
      start: index === 0 ? start : lastGap.end + 1,
      end: end < gapAhead.start ? end : gapAhead.start - 1,
    });
  });

  // Creating he last fragment after the last gap
  if (gaps.length > 0) {
    const veryLastGap = gaps[gaps.length - 1];
    // if the end position of the last gap is smaller than the feature's end position
    // TODO is thie 'if' redundant?
    if (veryLastGap.end < end) {
      fragments.push({
        ...feature,
        // Our starting point is the very end position of the very last gap
        start: veryLastGap.end + 1,
        end,
      });
    }
  }

  // console.log("--- gaps:", gaps);
  // console.log("--- frags:", fragments);

  // If no fragments were created, then simply return the original feature
  // without any modifications.
  // If fragments were created, add them to the feature and then return the
  // modified feature;
  return fragments.length === 0
    ? feature
    : {
        ...feature,
        locations: [
          {
            fragments,
          },
        ],
      };
};
