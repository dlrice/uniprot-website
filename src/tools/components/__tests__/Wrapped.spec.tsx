import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import renderWithRedux from '../../../shared/__test-helpers__/RenderWithRedux';

import Wrapped from '../Wrapped';
import { BLAST as mockData } from '../__mocks__/msaMocks.json';
import { getFullAlignmentLength } from '../../utils/sequences';
describe('Wrapped', () => {
  describe('BLAST', () => {
    let props;
    beforeEach(() => {
      props = {
        alignment: mockData,
        alignmentLength: mockData[0].length,
        conservationOptions: {},
        totalLength: getFullAlignmentLength(mockData, mockData[0].length),
        onMSAFeatureClick: jest.fn(),
        activeAnnotation: [],
        // activeAlignment={activeAlignment
        // highlightProperty={highlightProperty
        // annotation={annotation
        // activeId={mockData[1].accession
        // selectedMSAFeatures={selectedMSAFeatures
      };
    });

    it('should render', () => {
      const { asFragment } = renderWithRedux(<Wrapped {...props} />);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render with active sequence', () => {
      const { asFragment } = renderWithRedux(
        <Wrapped {...props} activeId={mockData[1].accession} />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
