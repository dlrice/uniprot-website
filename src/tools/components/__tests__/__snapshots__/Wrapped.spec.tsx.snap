// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrapped BLAST should render 1`] = `
<DocumentFragment>
  <div
    class="alignment-grid alignment-wrapped"
    data-testid="alignment-wrapped-view"
  />
</DocumentFragment>
`;

exports[`Wrapped BLAST should render with active sequence 1`] = `
<DocumentFragment>
  <div
    class="alignment-grid alignment-wrapped"
    data-testid="alignment-wrapped-view"
  />
</DocumentFragment>
`;
